src/test/java/picocli/CommandSubcommandFilterTest.java

```diff
@@ -0,0 +1,149 @@
+package picocli;
+
+import org.junit.*;
+import org.junit.contrib.java.lang.system.ProvideSystemProperty;
+import org.junit.contrib.java.lang.system.RestoreSystemProperties;
+import org.junit.contrib.java.lang.system.SystemErrRule;
+import org.junit.contrib.java.lang.system.SystemOutRule;
+import org.junit.rules.TestRule;
+import org.junit.Test;
+import java.util.Arrays;
+import java.util.concurrent.Callable;
+
+import static org.junit.Assert.*;
+
+public class CommandSubcommandFilterTest {
+    // allows tests to set any kind of properties they like, without having to individually roll them back
+    @Rule
+    public final TestRule restoreSystemProperties = new RestoreSystemProperties();
+
+    @Rule
+    public final ProvideSystemProperty ansiOFF = new ProvideSystemProperty("picocli.ansi", "false");
+
+    @Rule
+    public final SystemErrRule systemErrRule = new SystemErrRule().enableLog().muteForSuccessfulTests();
+
+    @Rule
+    public final SystemOutRule systemOutRule = new SystemOutRule().enableLog().muteForSuccessfulTests();
+
+    @CommandLine.Command(name = "aa")
+    private static class AA implements Callable<Integer>
+    {
+        public Integer call() throws Exception
+        {
+            return 0;
+        }
+    }
+
+
+    @CommandLine.Command(name = "c")
+    private static class C implements Callable<Integer>
+    {
+        public Integer call() throws Exception
+        {
+            return 0;
+        }
+    }
+
+    @CommandLine.Command(name = "b")
+    private static class B implements Callable<Integer>
+    {
+        public Integer call() throws Exception
+        {
+            return 0;
+        }
+    }
+
+
+    @CommandLine.Command(name = "a", subcommands = {AA.class}, preprocessor = A.AFilter.class)
+    private static class A implements Callable<Integer>
+    {
+
+        private static class AFilter implements CommandLine.IPreprocessor
+        {
+            public CommandLine.Model.CommandSpec preprocess(CommandLine.Model.CommandSpec commandSpec)
+            {
+                // verify it's context aware
+                assertEquals(2, commandSpec.subcommands().size());
+
+                if(Boolean.valueOf(System.getProperty("disable_aa"))) commandSpec.removeSubcommand("aa");
```

Comment:
The code style I use is space between the `if` and the opening bracket `(`, and redundant braces for the statement following the conditional.